enhancements/ingress/set_eip_nlb_ingress.md

```diff
@@ -0,0 +1,1319 @@
+
+### Non-Goals
+
+- Creation of EIPs in AWS.
+- Keep a track of changes in the EIP in the AWS environment.
+- Static IP usage with NLBs for API server.
+- Check the number of public subnets available and then compare number of subnets
+  with the number of EIPs the user provides.
+
+## Proposal
+
+EIP = AWS Elastic IP, NLB = AWS Network LoadBalancer 
+
+In this enhancement we would be adding API fields in the installer and the Ingress Controller CR
+to set the AWS EIP for the NLB Ingress Controller. The cluster ingress operator will then create 
+the ingress controller and get the EIP assigned to the service LoadBalancer with the annotation 
+`service.beta.kubernetes.io/aws-load-balancer-eip-allocations`
+Note: `service.beta.kubernetes.io/aws-load-balancer-eip-allocations` is a [kubernetes service annotation](https://kubernetes.io/docs/reference/labels-annotations-taints/#service-beta-kubernetes-io-aws-load-balancer-eip-allocations).
+
+
+### Workflow Description
+
+Pre-requisites: One must [allocate Elastic IP address](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/elastic-ip-addresses-eip.html#using-instance-addressing-eips-allocating)
+in their AWS environment which could be then used in the `eipAllocations` field of the IngressController CR.
+Note: The AWS account has some cap on the number of EIPs per region. By default, all AWS accounts have a quota of five (5) Elastic IP addresses per Region, because public (IPv4) 
+internet addresses are a scarce public resource. For more details please check [Elastic IP address quota](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/elastic-ip-addresses-eip.html#using-instance-addressing-limit)
+When this cap is hit you will get the following error: `Elastic IP address could not be allocated.
+The maximum number of addresses has been reached.`
+
+**cluster administrator** is a human user responsible for deploying a
+cluster. A cluster administrator also has the rights to modify the cluster level components.
+
+#### Set EIP using an installer:
+1. Cluster administrator creates `install-config.yaml` and adds the following in it. 
+```yaml
+apiVersion: v1
+baseDomain: example.com
+metadata:
+  name: test-cluster
+platform:
+  aws:
+    region: us-west-2
+    lbType: NLB
+      networkLoadBalancer:
+        eipAllocations:
+        - eipalloc-1111
+        - eipalloc-2222
+        - eipalloc-3333
+        - eipalloc-4444
+        - eipalloc-5555 
+pullSecret: '{"auths": ...}'
+sshKey: ssh-ed25519 AAAA...
+```  
+Here, the administrator needs to know how many public subnets the VPC ID of their cluster has.
+They may check it by clicking the `Load Balancers` tab then under the details click `VPC ID` under VPC, then click the link under `Main route table` and then check the column `Explicit subnet associations` which
```

Comment:
actually it is VPC tagged  with cluster name

Reply:
>>The steps listed here are assuming you are bringing your own VPC, that's already created. I think that needs to be stated here.

OK I will state it.

>>  Shouldn't we also have a workflow for when the installer creates the VPC and public subnets? In that case, you'd need the cluster admin to find how many AZs are in the region they are installing into and express that the number of public subnets == the number of AZs in a region

I need to think on this. Lets have a call to discuss this. I did not get this `express that the number of public subnets == the number of AZs in a region`

Reply:
Yes to be clear, I mean when the openshift-installer creates the VPC and subnets (the default functionality when `platform.aws.subnets` are NOT specified in install-config). SO this is NOT the BYO VPC use case.

For installer-created-VPCs, how does a cluster admin accurately predict how many EIPAllocations to bring so their cluster doesn't fail to install? Maybe they don't know what public subnets will be created by the openshift-installer. 

I think it would be good to have a workflow for this. To illustrate further, the workflow could start like:

1. Cluster admin wants to use EIPs, but isn't bringing their own VPC, so they need to know exactly how many EIPs to create.
2. Cluster admin should review https://aws.amazon.com/about-aws/global-infrastructure/regions_az/ to find out how many AZs are in the region they are installing into.
3. Cluster admin should use the same number of EIPs as AZs in their region.

Let me know if that makes sense, or if anyone has a reason why this might be redundant. Cluster admins probably know how many AZs they are installing into, but I suppose a workflow like this would provide a concrete example of how this would happen if they don't.

> express that the number of public subnets == the number of AZs in a region

Sorry, I think this is a typo, I meant "express that the number of ~~public subnets~~ EIPs == the number of AZs in a region`